### PR TITLE
adding support forWatermark in querybuilder

### LINKED ISSUE #1564: Support forWatermark in QueryBuilder
### Similar requests

- [X] I have searched and found no existing similar requests

### What problems are you trying to solve?

We want to enable function `meta::datalake::functions::forWatermark<T>(collection: T[*], watermak: String[1]): T{*]` in query builder.

The feature should understand `forWatermark` function and let users set the value of the watermark as either a constant value or a parameter value. 

Sample Queries To Support:
`{watermark: String[1]| model::Person.all()->forWatermark($watermark)}`
`{| model::Person.all()->forWatermark('testWatermark')}`

### Describe the solution you would like to see implemented

Steps

- Decide where to put access point of toggling watermark function. i.e advanced options? tds  main buttons ? 
- Allow toggling of function. If function is `on` allow users to Drag and drop parameters (could reuse derived properties editor as similar in that its a function which requires inputs) to set values or use constant values
- add processing state builder/value spec builders for `forWatermark` function

### PR DESCRIPTION
**Pull request by xannem**
fixes #1564 

## Summary
<hr/> 
This feature supports the `forWatermark` function in Query builder. 
In the query, there will now be a quick description on the top left of the query panel with icons that provide a quick summary of any unique qualities of the query. The waterdrop icon represents the watermark and will appear when a watermark is present. Users can choose to add and edit their watermarks by clicking this icon, or by clicking 'edit watermark' from the advanced options in the right side of the top panel. <br/><br/>

Kapture 2022-10-26 at 14 03 56

Users can drag and drop parameters from their parameter list as the watermark value, or type in their own constant. 
<br/><br/>

Kapture 2022-10-24 at 17 42 36

- [x] Decide where to put access point of toggling watermark function. i.e advanced options? tds main buttons ?
- [x] Allow toggling of function. If function is on allow users to Drag and drop parameters (could reuse derived properties editor as similar in that its a function which requires inputs) to set values or use constant values
- [x] add processing state builder/value spec builders for forWatermark function

<hr/> 

MISC Patch info: 
- [x] Add `Modal` component

## How did you test this change?

- [ ] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

### CODE REVIEW

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+);
+
+export const QueryBuilderWatermarkEditor = observer(
+  (props: { queryBuilderState: QueryBuilderState }) => {
+    const { queryBuilderState } = props;
+
+    const watermarkState = queryBuilderState.watermarkState;
+
+    const handleClose = (): void => {
+      if (watermarkState.value instanceof PrimitiveInstanceValue) {
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:143]: why are we doing this logic in closing of the modal, arent we updating the value spec in real time when the user is editing the value spec editor?

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+import { VariableExpressionViewer } from '../QueryBuilderParametersPanel.js';
+import { BasicValueSpecificationEditor } from '../shared/BasicValueSpecificationEditor.js';
+
+const WatermarkValueEditor = observer(
+  (props: {
+    watermarkState: QueryBuilderWatermarkState;
+    queryBuilderState: QueryBuilderState;
+  }) => {
+    const { watermarkState } = props;
+    const variable = watermarkState.value as VariableExpression;
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:56]: why are we making this casting here, we can't assume all value specs will be Variable Expressions

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+      QueryBuilderParameterDragSource,
+      void,
+      { isParameterValueDragOver: boolean }
+    >(
+      () => ({
+        accept: [QUERY_BUILDER_PARAMETER_DND_TYPE],
+        drop: (item, monitor): void => {
+          const itemType = item.variable.parameter.genericType?.value.rawType;
+          if (
+            !monitor.didDrop() &&
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:77]: Both the filtering and dropping logic should match what the `forWatermark` function accepts. It accepts a parameter of type `String` with multiplicity of `1..1`. 
You shouldn't need the current valueSpec type.

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        }),
+      }),
+      [handleDrop],
+    );
+
+    return (
+      <>
+        {watermarkState.value && (
+          <div key={variable.name} className="panel__content__form__section">
+            <div className="panel__content__form__section__header__label">
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:100]: This div should not be necessary. The value spec editor should handle rendering value spec of type variable expression.

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+                  obseverContext={
+                    watermarkState.queryBuilderState.observableContext
+                  }
+                  typeCheckOption={{
+                    expectedType: graph.getPrimitiveType(PRIMITIVE_TYPE.STRING),
+                    match:
+                      parameterType ===
+                      graph.getPrimitiveType(PRIMITIVE_TYPE.STRING),
+                  }}
+                  resetValue={(): void => watermarkState.setValue(undefined)}
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:125]: I think resetValue should probably reset the value spec to the default value which I think we can make `watermarkValue`

File: packages/legend-query-builder/src/stores/QueryBuilderValueSpecificationBuilder.ts [hunk 6]
```diff
@@ -145,6 +146,15 @@ export const buildLambdaFunction = (
     lambdaFunction.expressionSequence[0] = filterFunction;
   }
 
+  // build watermark
+  const watermarkFunction = buildWatermarkExpression(
+    queryBuilderState.watermarkState,
+    getAllFunction,
```
Comment (YannanGao-gs) [inline packages/legend-query-builder/src/stores/QueryBuilderValueSpecificationBuilder.ts:152]: Filter will be replaced

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+import {
+  type QueryBuilderParameterDragSource,
+  QUERY_BUILDER_PARAMETER_DND_TYPE,
+} from '../../stores/QueryBuilderParametersState.js';
+import type { QueryBuilderState } from '../../stores/QueryBuilderState.js';
+import type { QueryBuilderWatermarkState } from '../../stores/watermark/QueryBuilderWatermarkState.js';
+import { VariableExpressionViewer } from '../QueryBuilderParametersPanel.js';
+import { BasicValueSpecificationEditor } from '../shared/BasicValueSpecificationEditor.js';
+
+const WatermarkValueEditor = observer(
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:50]: can we isolate this component to assume the watermark value is defined. The modal component will handle the interaction of whether its defined or not. I would just make this component take as inputs
`watermarkValue: ValueSpec.., queryBuilderState: QueryBuilderState`

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+                <>
+                  <WatermarkValueEditor
+                    watermarkState={watermarkState}
+                    queryBuilderState={queryBuilderState}
+                  />
+                  <PanelDivider />
+                  <PanelFormListItems title="List of available parameters">
+                    {queryBuilderState.parametersState.parameterStates
+                      .length === 0 && <> No parameters in query </>}
+                    {queryBuilderState.parametersState.parameterStates.map(
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:185]: As mentioned, lets isolate the parameters we show as only parameters that meet the input of the `forWatermark` function. 
`meta::datalake::functions::forWatermark<T>(collection: T[*], watermak: String[1]): T{*]` as seen here the watermark value is of type string with multiplicity 1.

File: packages/legend-query-builder/src/stores/__tests__/QueryBuilderBuildLambdaRoundtrip.test.tsx [hunk 9]
```diff
@@ -273,6 +281,19 @@ const cases: RoundtripTestCase[] = [
...
+    TEST_DATA_lambda_watermark_Constant,
+    undefined,
+  ],
+  [
+    'Watermark with parameter',
+    forWatermarkCtx,
+    TEST_DATA_lambda_watermark_Parameter,
+    undefined,
+  ],
   [
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/stores/__tests__/QueryBuilderBuildLambdaRoundtrip.test.tsx:297]: lets add a test for having query logic after the waterMark. 
so forWatermark()->project() , forWatermark()->filter().

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]

File: packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkState.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  toogleWatermark(): void {
+    if (this.value) {
+      this.setValue(undefined);
+    } else {
+      const watermarkConstant = new PrimitiveInstanceValue(
+        GenericTypeExplicitReference.create(
+          new GenericType(
+            this.queryBuilderState.graphManagerState.graph.getPrimitiveType(
```
Comment (YannanGao-gs) [inline packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkState.ts:53]: i think maybe you could set the value here to be '', which means by default we'll generate ->forWatermark('')

File: packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkStateBuilder.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    ]),
+    `Can't process watermark() expression: only support watermark() immediately following either getAll() or filter()`,
+  );
+
+  QueryBuilderValueSpecificationProcessor.process(
+    precedingExpression,
+    queryBuilderState,
+  );
+
+  const watermarkLambda = expression.parametersValues[1];
```
Comment (YannanGao-gs) [inline packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkStateBuilder.ts:50]: is it a `lambda`?  maybe `watermarkValue`

File: packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkValueSpecificationBuilder.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  SimpleFunctionExpression,
+  extractElementNameFromPath,
+  type VariableExpression,
+} from '@finos/legend-graph';
+import { QUERY_BUILDER_SUPPORTED_FUNCTIONS } from '../../graphManager/QueryBuilderSupportedFunctions.js';
+import type { QueryBuilderWatermarkState } from './QueryBuilderWatermarkState.js';
+
+export const buildWatermarkExpression = (
+  watermarkState: QueryBuilderWatermarkState,
+  getAllFunc: SimpleFunctionExpression,
```
Comment (YannanGao-gs) [inline packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkValueSpecificationBuilder.ts:27]: is filter before forwaterMark or after? see one of the asserstions in QueryBuilderWatermarkStateBuilder.ts

File: packages/legend-art/src/components/dialog/Modal.tsx [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+};
+
+export const ModalTitle: React.FC<{
+  title: string;
+  icon?: React.ReactNode;
+  className?: string;
+}> = (props) => {
+  const { title, icon } = props;
+  return (
+    <div className="modal__title">
```
Comment (YannanGao-gs) [inline packages/legend-art/src/components/dialog/Modal.tsx:41]: add className ?

File: packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkStateBuilder.ts (hunk 11 above)
Comment (xannem) [reply packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkStateBuilder.ts:50]: true! 🙇

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx (hunk 5 above)
Comment (xannem) [reply packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:125]: Everytime the value spec editor is changed though the value becomes the watermark value, so if I changed it to `watermarkState.setValue(watermarkState.value)` I believe a difference never occurs? 🤔

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx (hunk 8 above)
Comment (xannem) [reply packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:185]: makes sense! 🙇

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx (hunk 7 above)
Comment (xannem) [reply packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:50]: I believe it needs the watermarkState still to call `setValue` on handleDrop but I can def remove the logic of defined or not and use the `watermarkValue` as input 👍

File: packages/legend-query-builder/src/stores/QueryBuilderValueSpecificationBuilder.ts (hunk 6 above)
Comment (xannem) [reply packages/legend-query-builder/src/stores/QueryBuilderValueSpecificationBuilder.ts:143]: thank you!

File: packages/legend-query-builder/src/stores/__tests__/QueryBuilderBuildLambdaRoundtrip.test.tsx [hunk 14]
```diff
@@ -273,6 +283,32 @@ const cases: RoundtripTestCase[] = [
...
+  ],
+  [
+    'Watermark on filter',
+    forWatermarkCtx,
+    TEST_DATA_lambda_watermark_filter_Constant,
+    undefined,
+  ],
+  [
+    'Watermark on olap',
+    forWatermarkCtx,
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/stores/__tests__/QueryBuilderBuildLambdaRoundtrip.test.tsx:307]: Watermark with project() and olapGroupBy()

File: packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkStateBuilder.ts [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+  const precedingExpression = guaranteeType(
+    expression.parametersValues[0],
+    SimpleFunctionExpression,
+    `Can't process forWatermark() expression: only support forWatermark() immediately following an expression`,
+  );
+
+  assertTrue(
+    matchFunctionName(precedingExpression.functionName, [
+      QUERY_BUILDER_SUPPORTED_FUNCTIONS.GET_ALL,
+      QUERY_BUILDER_SUPPORTED_FUNCTIONS.FILTER,
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkStateBuilder.ts:40]: Don't we only support all() -> forWatermark(). Don't think we support getAll()->filter()->forWatermark()

File: packages/legend-art/src/components/dialog/Modal.tsx [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+export const ModalHeader: React.FC<{
+  title?: string;
+  children?: React.ReactNode;
+  className?: string;
+}> = (props) => {
+  const { title, children } = props;
+
+  if (!children && title) {
+    return (
+      <div className="modal__header">
```
Comment (YannanGao-gs) [inline packages/legend-art/src/components/dialog/Modal.tsx:57]: + className ?

File: packages/legend-query-builder/src/graphManager/protocol/pure/v1/V1_QueryValueSpecificationBuilderHelper.ts [hunk 17]
```diff
@@ -570,6 +570,75 @@ export const V1_buildGroupByFunctionExpression = (
...
+    (parameters[1] as V1_ValueSpecification).accept_ValueSpecificationVisitor(
+      new V1_ValueSpecificationBuilder(
+        compileContext,
+        processingContext,
+        openVariables,
+      ),
+    ),
+  ];
+
+  assertTrue(
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/graphManager/protocol/pure/v1/V1_QueryValueSpecificationBuilderHelper.ts:626]:
```
const watermarkValueParam = guarnatNonNullable(processedParam[1]);
const watermarkValueParamType = watermarkValueParam.genericType?.value.rawType;
assertTrue(
  compileContext.graph.getPrimitiveType(PRIMITE_TYPE.STRING) === watermarkValueParamType,
  `Can't build forWatermark() expression: watermark parameter is expected to be of type string. 
)

File: packages/legend-query-builder/src/components/QueryBuilder.tsx [hunk 18]
```diff
@@ -237,131 +243,154 @@ export const QueryBuilder = observer(
...
-                      {!queryBuilderState.isParameterSupportDisabled && (
+              <div className="query-builder__sub-header__content__icons">
+                {queryBuilderState.watermarkState.value && (
+                  <button
+                    className="panel__header__action"
+                    onClick={openWatermark}
+                    tabIndex={-1}
+                    title="Edit Watermark"
+                  >
+                    <WaterDropIcon />
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/QueryBuilder.tsx:262]: Should we have maybe an indicator on the icon if the `forWatermark` function is on for the query. 
Maybe have the icon be a different color if the function is on.

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+        </div>
+      </PanelFormSection>
+    );
+  },
+);
+
+export const QueryBuilderWatermarkEditor = observer(
+  (props: { queryBuilderState: QueryBuilderState }) => {
+    const { queryBuilderState } = props;
+
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:128]: remove empty line

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+    const watermarkState = queryBuilderState.watermarkState;
+
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:130]: remove empty line

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+                    watermarkState={watermarkState}
+                    watermarkValue={watermarkState.value}
+                  />
+                  <PanelDivider />
+                  <PanelFormListItems title="List of available parameters">
+                    {queryBuilderState.parametersState.parameterStates
+                      .length === 0 && <> No available parameters in query </>}
+                    {queryBuilderState.parametersState.parameterStates
+                      .filter((pState) => {
+                        const isMuliplicityOne =
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:167]: lets move this out to the top as a helper method.
`isParamaterCompatibleWithWaterMark`
for the string comparison should be 
queryBuilder.graphManagerState.gaph.getPrimitiveType(PRIMITIVE_STRING).
Also for the multiplicity comparator use `multiplicityComparator` that Yannan added

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]
few comments. looks good.

File: .changeset/calm-brooms-cough.md [hunk 22]
```diff
@@ -0,0 +1,5 @@
+---
+'@finos/legend-query-builder': minor
+---
+
+add support for function `forWatermark` in QueryBuilder [#1564](https://github.com/finos/legend-studio/issues/1564)
```
Comment (akphi) [inline .changeset/calm-brooms-cough.md:5]:
```suggestion
Add support for function `forWatermark()` [#1564](https://github.com/finos/legend-studio/issues/1564)
```

File: fixtures/legend-mock-server/src/index.ts [hunk 23]
```diff
@@ -15,9 +15,9 @@
  */
 
 import { fastify } from 'fastify';
-import { fastifyCors } from '@fastify/cors';
 import TAXONOMY_TREE_DATA from './TEST_DATA__TaxonomyTreeData.json' assert { type: 'json' };
 import DOCUMENTATION_DATA from './DummyDocumentationData.json' assert { type: 'json' };
+import fastifyCors from '@fastify/cors';
```
Comment (akphi) [inline fixtures/legend-mock-server/src/index.ts:20]: why? Please revert this change

File: packages/legend-art/src/components/panel/Panel.tsx [hunk 24]
```diff
@@ -228,3 +262,40 @@ export const PanelFormBooleanField = observer(
...
+        </div>
+      )}
+      <div className="panel__content__form__section__list__items">
+        {children}
+      </div>
+    </PanelFormSection>
+  );
+};
+
+export const PanelFormActions: React.FC<{
```
Comment (akphi) [inline packages/legend-art/src/components/panel/Panel.tsx:291]: let's use `generateSimpleDIVComponent`

File: packages/legend-art/style/base/_input.scss [hunk 25]
```diff
@@ -34,6 +34,10 @@
   }
 }
 
+.input__small {
```
Comment (akphi) [inline packages/legend-art/style/base/_input.scss:37]: --> `input--small`

File: packages/legend-art/style/base/_modal.scss [hunk 26]
```diff
@@ -55,7 +55,7 @@
     @include flexVCenter;
   }
 
-  &__header__action {
+  &__header__action-btn {
```
Comment (akphi) [inline packages/legend-art/style/base/_modal.scss:58]: let's revert this change, action sometimes doesn't have to be button.

File: packages/legend-query-builder/src/components/shared/BasicValueSpecificationEditor.tsx [hunk 27]
```diff
@@ -196,7 +196,10 @@ const StringPrimitiveInstanceValueEditor = observer(
     } = props;
     const useSelector = Boolean(selectorConfig);
     const applicationStore = useApplicationStore();
-    const value = valueSpecification.values[0] as string;
+    const value =
```
Comment (akphi) [inline packages/legend-query-builder/src/components/shared/BasicValueSpecificationEditor.tsx:199]: when we do see the case where this returns `undefined` (i.e. has less than 1 item)? if we're just trying to get around TS here, we can use `guaranteeIsString(...)`

File: packages/legend-query-builder/src/stores/QueryBuilderState.ts [hunk 28]
```diff
@@ -116,11 +119,13 @@ export abstract class QueryBuilderState implements CommandRegistrar {
       functionsExplorerState: observable,
       fetchStructureState: observable,
       filterState: observable,
+      watermarkState: observable,
       resultState: observable,
       textEditorState: observable,
       unsupportedQueryState: observable,
       showFunctionsExplorerPanel: observable,
       showParametersPanel: observable,
+      isEditingWatermark: observable,
```
Comment (akphi) [inline packages/legend-query-builder/src/stores/QueryBuilderState.ts:128]: let's put these in `watermarkState` to keep `QueryBuilderState` clean

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 29]
```diff
@@ -0,0 +1,209 @@
...
+                          pState.parameter.multiplicity ===
+                          queryBuilderState.graphManagerState.graph.getTypicalMultiplicity(
+                            TYPICAL_MULTIPLICITY_TYPE.ONE,
+                          );
+
+                        const isString =
+                          pState.parameter.genericType?.value.rawType.name ===
+                          'String';
```
Comment (akphi) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx]: Please never do this, let's use the enum `PRIMITIVE_TYPE`

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 30]
```diff
@@ -0,0 +1,209 @@
...
+
+                        if (isString && isMuliplicityOne) {
```
Comment (akphi) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx]: --> `return isString && isMuliplicityOne`

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 31]
```diff
@@ -0,0 +1,209 @@
...
+                    watermarkValue={watermarkState.value}
+                  />
+                  <PanelDivider />
+                  <PanelFormListItems title="List of available parameters">
+                    {queryBuilderState.parametersState.parameterStates
+                      .length === 0 && <> No available parameters in query </>}
+                    {queryBuilderState.parametersState.parameterStates
+                      .filter((pState) => {
+                        const isMuliplicityOne =
+                          pState.parameter.multiplicity ===
```
Comment (akphi) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx]: I don't think this is a safe check, there's a method to do multiplicity equality, `multiplicityComparator`, let's use that

File: packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkStateBuilder.ts (hunk 15 above)
Comment (xannem) [reply packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkStateBuilder.ts:40]: yep!

**Review by akphi — CHANGES_REQUESTED** [review]

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 32]
```diff
@@ -0,0 +1,212 @@
...
+
+export const QueryBuilderWatermarkEditor = observer(
+  (props: { queryBuilderState: QueryBuilderState }) => {
+    const { queryBuilderState } = props;
+    const watermarkState = queryBuilderState.watermarkState;
+    const handleClose = (): void => {
+      queryBuilderState.setIsEditingWatermark(false);
+    };
+
+    const isParamaterCompatibleWithWaterMark = (
```
Comment (akphi) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:135]: let's move it out of component to keep the component leaner.

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 33]
```diff
@@ -0,0 +1,212 @@
...
+                  <WatermarkValueEditor
+                    watermarkState={watermarkState}
+                    watermarkValue={watermarkState.value}
+                  />
+                  <PanelDivider />
+                  <PanelFormListItems title="List of available parameters">
+                    {queryBuilderState.parametersState.parameterStates
+                      .length === 0 && <> No available parameters in query </>}
+                    {queryBuilderState.parametersState.parameterStates
+                      .filter((pState) => {
```
Comment (akphi) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:183]: --> `.filter(isParamaterCompatibleWithWaterMark)`

File: packages/legend-query-builder/src/components/QueryBuilder.tsx (hunk 18 above)
Comment (akphi) [reply packages/legend-query-builder/src/components/QueryBuilder.tsx:262]: I think so to reduce clutters. The `grey` part sounds pretty sad, but that's for now

File: packages/legend-art/style/base/_modal.scss (hunk 26 above)
Comment (xannem) [reply packages/legend-art/style/base/_modal.scss:58]: ! true

File: packages/legend-query-builder/src/stores/__tests__/QueryBuilderBuildLambdaRoundtrip.test.tsx [hunk 34]
```diff
@@ -273,6 +283,32 @@ const cases: RoundtripTestCase[] = [
...
+    undefined,
+  ],
+  [
+    'Watermark with parameter value',
+    forWatermarkCtx,
+    TEST_DATA_lambda_watermark_Parameter,
+    undefined,
+  ],
+  [
+    'Watermark with filter',
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/stores/__tests__/QueryBuilderBuildLambdaRoundtrip.test.tsx:300]: with filter()

File: packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkState.ts [hunk 35]
```diff
@@ -0,0 +1,107 @@
...
+      resetValue: action,
+      enableWatermark: action,
+      hashCode: computed,
+    });
+
+    this.queryBuilderState = queryBuilderState;
+  }
+
+  isParamaterCompatibleWithWaterMark = (
+    pState: LambdaParameterState,
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkState.ts:50]: lets have the input be `parameter: VariableExpression`

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 36]
```diff
@@ -0,0 +1,190 @@
...
+      { isParameterValueDragOver: boolean }
+    >(
+      () => ({
+        accept: [QUERY_BUILDER_PARAMETER_DND_TYPE],
+        drop: (item, monitor): void => {
+          const itemType = item.variable.parameter.genericType?.value.rawType;
+          if (
+            !monitor.didDrop() &&
+            // Only allows parameters with muliplicity 1 and type string
+            item.variable.parameter.multiplicity ===
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:76]: lets use the helper function in watermarkState to determine if this variable expression is compatible

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 37]
```diff
@@ -0,0 +1,190 @@
...
+
+    return (
+      <PanelFormSection>
+        <div className="query-builder__parameter-editor">
+          <PanelDropZone
+            isDragOver={isParameterValueDragOver}
+            dropTargetConnector={dropTargetConnector}
+          >
+            <BasicValueSpecificationEditor
+              valueSpecification={guaranteeNonNullable(watermarkValue)}
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:102]: you shouldnt need the guarantNonNullable if the input for this component is `watermarkValue: ValueSpecification`

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 38]
```diff
@@ -0,0 +1,190 @@
...
+                prompt="Enable watermark"
+                update={(): void => watermarkState.enableWatermark()}
+              />
+              {watermarkState.value && (
+                <>
+                  <WatermarkValueEditor
+                    watermarkState={watermarkState}
+                    watermarkValue={watermarkState.value}
+                  />
+                  <PanelDivider />
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:158]: can we create a const before the rendering such as 
```
const compatibleParameters = queryBuilderState.parametersState.parameterStates.map(p => p.parameter).filter(watermarkState.isParamaterCompatibleWithWaterMark)
```
then use that when rendering 
```
   {compatibleParameters
                      .length === 0 && <> No available parameters in query </>}
```
and 
```
compatibleParameters.map(<VariableExpressionViewer...>)

```

File: packages/legend-query-builder/src/stores/QueryBuilderState.ts (hunk 28 above)
Comment (MauricioUyaguari) [reply packages/legend-query-builder/src/stores/QueryBuilderState.ts:128]: + 1

File: packages/legend-query-builder/src/components/shared/BasicValueSpecificationEditor.tsx (hunk 27 above)
Comment (MauricioUyaguari) [reply packages/legend-query-builder/src/components/shared/BasicValueSpecificationEditor.tsx:199]: i think we should go back to what it was before. looking at the code we added this because we weren't populating the watermark string value.

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx (hunk 37 above)
Comment (xannem) [reply packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:102]: oh makes sense!

File: packages/legend-query-builder/src/components/shared/BasicValueSpecificationEditor.tsx [hunk 39]
```diff
@@ -196,7 +196,10 @@ const StringPrimitiveInstanceValueEditor = observer(
...
+      valueSpecification.values.length > 0
+        ? (valueSpecification.values[0] as string)
+        : '';
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/shared/BasicValueSpecificationEditor.tsx:202]: Can we revert this to the way it was. I don't think this is needed anymore

File: packages/legend-query-builder/src/stores/QueryBuilderState.ts [hunk 40]
```diff
@@ -97,6 +99,7 @@ export abstract class QueryBuilderState implements CommandRegistrar {
   queryCompileState = ActionState.create();
   showFunctionsExplorerPanel = false;
   showParametersPanel = false;
+  isEditingWatermark = false;
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/stores/QueryBuilderState.ts:102]: As An suggested, lets move this to watermarkState

File: .changeset/calm-brooms-cough.md [hunk 41]
```diff
@@ -0,0 +1,5 @@
+---
+'@finos/legend-query-builder': minor
+---
+
+Add support for function `forWatermark()` [#1564](https://github.com/finos/legend-studio/issues/1564)
```
Comment (akphi) [inline .changeset/calm-brooms-cough.md:5]:
```suggestion
Add support for function `forWatermark()` ([#1564](https://github.com/finos/legend-studio/issues/1564))
```

File: packages/legend-art/style/base/_input.scss [hunk 42]
```diff
@@ -34,6 +34,10 @@
   }
 }
 
+.input--small {
```
Comment (akphi) [inline packages/legend-art/style/base/_input.scss:37]: I think this is something we put in `panel.scss` instead of here. This small input width is really just a panel thing

File: packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkState.ts [hunk 43]
```diff
@@ -0,0 +1,114 @@
...
+  multiplicityComparator,
+  type VariableExpression,
+} from '@finos/legend-graph';
+import { type Hashable, hashArray } from '@finos/legend-shared';
+import { makeObservable, observable, action, computed } from 'mobx';
+import { QUERY_BUILDER_HASH_STRUCTURE } from '../../graphManager/QueryBuilderHashUtils.js';
+import type { QueryBuilderState } from '../QueryBuilderState.js';
+
+export class QueryBuilderWatermarkState implements Hashable {
+  queryBuilderState: QueryBuilderState;
```
Comment (akphi) [inline packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkState.ts:34]: --> `readonly`

File: packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx [hunk 44]
```diff
@@ -0,0 +1,185 @@
...
+                  <PanelFormListItems title="List of available parameters">
+                    {compatibleParameters.length === 0 && (
+                      <> No available parameters in query </>
```
Comment (akphi) [inline packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx:157]: --> `No available paramerters` is sufficient

**Review by akphi — CHANGES_REQUESTED** [review]
Looking great! only minor comments left

File: packages/legend-art/style/base/_input.scss (hunk 42 above)
Comment (xannem) [reply packages/legend-art/style/base/_input.scss:37]: true!

File: packages/legend-art/style/base/_panel.scss [hunk 45]
```diff
@@ -185,6 +185,10 @@
   }
 }
 
+.input--small {
```
Comment (akphi) [inline packages/legend-art/style/base/_panel.scss:188]: We don't want this global class, let's put it at appropriate scope

**Review by akphi — APPROVED** [review]

**Review by MauricioUyaguari — APPROVED** [review]
